scipy/sparse/linalg/_dsolve/linsolve.py

```diff
@@ -683,63 +703,40 @@ def spsolve_triangular(A, b, lower=True, overwrite_A=False, overwrite_b=False,
     if b.ndim not in [1, 2]:
         raise ValueError(
             f'b must have 1 or 2 dims but its shape is {b.shape}.')
-    if A.shape[0] != b.shape[0]:
+    if M != b.shape[0]:
         raise ValueError(
             'The size of the dimensions of A must be equal to '
             'the size of the first dimension of b but the shape of A is '
             f'{A.shape} and the shape of b is {b.shape}.'
         )
 
-    # Init x as (a copy of) b.
-    x_dtype = np.result_type(A.data, b, np.float64)
-    if overwrite_b:
-        if np.can_cast(b.dtype, x_dtype, casting='same_kind'):
-            x = b
-        else:
-            raise ValueError(
-                f'Cannot overwrite b (dtype {b.dtype}) with result '
-                f'of type {x_dtype}.'
-            )
-    else:
-        x = b.astype(x_dtype, copy=True)
+    A = A._asfptype()  # upcast to a floating point format
```

Comment:
Ideally we could avoid this call, as it's a private method, but also because it can invoke `astype()` followed by another `astype()` in the lines below.

If we determine the type that `_asfptype()` would have resulted in without doing the cast, then we can pass that to `np.promote_types()` below directly.

Reply:
Thanks, I've replaced
```diff
-    A = A._asfptype()  # upcast to a floating point format
-    result_dtype = np.promote_types(A.dtype, b.dtype)
+    result_dtype = np.promote_types(np.promote_types(A.dtype, np.float32), b.dtype)
```
in 2df1a223b24f1c87b20422813864caeea9cf1068, hoping that this also has the intended effect. Do you agree with this solution? Alternatively, I could create a new function `_fptype` that returns the type that `_asfptype` would cast into.